Grenade fixes / Updates

Kell-E opened the pull request:
Currently you can shove any size beaker into any size grenades, this allows you to shove gigantic beakers into tiny grenades and create round ending explosions with minimal ease.

Small grenades (The basic, baby ones) can take any normal sized beaker.

Large grenades (Chemical grenades) can take any large beaker.

Also I've converted the EMP / Smoke Grenade and Flashbangs over to use chemicals instead of being their own unique type of grenades. 

It's worth noting that this is not live server friendly, it turns flashbangs into an instant fight ender regardless of your protection. I'll be updating flash powder along side this PR so that it actually cares about the reagent volume and properly checks for protection. 
This is fixed by #31061 

🆑 Yvesza
balance: Small grenades can now only take normal sized beakers
balance: Large grenades can take small or large beakers
balance: Smoke grenades will no longer cause you to cough and drop held objects
balance: Flashbangs affect a smaller radius but have a more concentrated effect in the very center of their blast
tweak: Smoke grenades and EMP grenades now use grenade casings
tweak: Tear gas grenades are now properly counted as being large chemical grenades
tweak: Pre-built grenades now default to a 3 second timer, instead of 5
tweak: Uranium no longer spreads goop on contact with turf
🆑

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -102,6 +103,8 @@
 				to_chat(user, "<span class='warning'>The grenade can not hold more containers.</span>")
 				return
+			if(is_type_in_list(W, disallowed_containers))
+				to_chat(user, "<span class='warning'>The grenade is too small for the [W].</span>")
```
`SPAN_*` defines. Use `\the` if prefacing vars with `the`. Also, use `src` instead of assuming it will always be 'grenade'. You never know what a child object or adminbussed name might end up being.

```suggestion
				to_chat(user, SPAN_WARNING("\The [src] is too small for \the [W]."))
```

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -324,3 +328,65 @@
+
+/obj/item/grenade/chem_grenade/flashbang
+	name = "Flashbang"
+	desc = "A grenade designed to blind, stun and disorient by means of an extremely bright flash and loud explosion."
```
```suggestion
	desc = "A grenade designed to blind, stun, and disorient by means of an extremely bright flash and a loud explosion."
```

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -324,3 +328,65 @@
+
+	beakers += B1
+	beakers += B2
+	icon_state = initial(icon_state) +"_locked"
```
Is there a reason you're not using `update_icon()` here instead? All `icon_state` operations should be in update icon if possible.

Kell-E replied:
I wasn't aware of the way that we were intended to handle icon_state changes, I've updated the code in a few places to reflect that!

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -324,3 +328,65 @@
+
+	beakers += B1
+	beakers += B2
+	icon_state = initial(icon_state) +"_locked"
```
As above re update_icon

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -324,3 +328,65 @@
+
+	beakers += B1
+	beakers += B2
+	icon_state = initial(icon_state) +"_locked"
```
As above re update_icon

SierraKomodo reviewed (changes requested):
Also did a quick round of testing with the grenades:

- The EMP grenade spawns glowing goo instead of an EMP.
- The flashbang spawns a spark instead of an actual flashbang.
- Not going to be blocking, but this makes me sad: The grenades have lost their unique identifying sprites with these changes.

Kell-E commented:
I'll see what I can do to preserve the unique look that EMP's have

Kell-E commented:
During testing I discovered a bug with flash powder, I had to PR #31072 as a fix otherwise this is entirely not safe for live as flashbangs will down entire rooms of people for silly lengths of time protection or not.

The goop still happens with EMP, it's a product of the way that grenades mix and spread reagents and the only (easy) way to fix it is to remove the goop creation from uranium.

SierraKomodo commented on code/game/objects/items/weapons/grenades/chem_grenade.dm:
```diff
@@ -13,6 +13,7 @@
 	var/obj/item/device/assembly_holder/detonator = null
 	var/list/beakers = new/list()
 	var/list/allowed_containers = list(/obj/item/reagent_containers/glass/beaker, /obj/item/reagent_containers/glass/bottle)
+	var/list/disallowed_containers = list(/obj/item/reagent_containers/glass/beaker/bluespace, /obj/item/reagent_containers/glass/beaker/large, /obj/item/reagent_containers/glass/beaker/insulated/large, /obj/item/reagent_containers/glass/beaker/bowl)
```
```suggestion
	var/list/disallowed_containers = list(
		/obj/item/reagent_containers/glass/beaker/bluespace,
		/obj/item/reagent_containers/glass/beaker/large,
		/obj/item/reagent_containers/glass/beaker/insulated/large,
		/obj/item/reagent_containers/glass/beaker/bowl
	)
```

SierraKomodo reviewed (commented):
Will do another test run once pain stuff kicks in.

Kell-E committed:
Grenade fixes

Currently you can shove any size beaker into grenades, including some things that count as beakers and can contain a hilariously large reagent count to create round ending grenades.

Small grenades (The basic, baby ones) can take any normal sized beaker.

Large grenades (Chemical grenades) can take any large beaker.

Also I've converted the EMP / Smoke Grenade and Flashbangs over to use chemicals instead of being their own unique type of grenades. 

It's worth noting that this massively overbuffs flashbangs because flash powder is in desperate need of a retouch, which I'll touch up shortly.
Fixes based upon feedback


Minor tweaks plus bugfix

Reduces the base timer on igniter assemblies to three from five. Fixes the incorrect EMP recipe. 
Uranium goop removal


Changes based upon review + Bowl fix on large grenades


Minor text tweak

a

Spookerton commented:
strong dislike of `generalizing` things onto other systems for the sake of it. The only thing loose inheritance of behavior achieves is needing to spend extra work working around or undoing this kind of thing when something further down the new tree changes the entire game balance.

Kell-E commented:
I can cut the generalization out of this and reduce it down to the grenade limitations if that's requested, just say the word!

Spookerton reviewed (dismissed):
bluh

SierraKomodo reviewed (changes requested):
This seems to effectively nerf flashbangs into doing nothing useful. They cause a brief flash of white, but no deafness and no stunning - Basically the same thing that clicking a flash on itself does.

It may be worth undoing the generalization of these special grenades, between Spook's concerns and things like the flashbang just not working. Also, this appears to have also removed the ability for flashbangs to damage blobs.

Kell-E commented:
Alrighty! I'll take a look over this, I didn't even realize flashbangs damaged blobs. I'll remove the generalization soon!
